### PR TITLE
Tests for do_updatedb (Issue 13944), alternate suggestion

### PR DESCRIPTION
**Pull request by arminfuerst**
This is an alternate suggestion to PR #14099 for introducing tests for do_updatedb.
In the first PR I pulled everything needed for the tests out of apps.c towards the tests, because there are functions in apps.c conflicting with the testing framework.
In this PR I tried the opposite approach: I included apps.c completely in the tests and pushed the conflicting functions (which turned out to be only one...) out of apps.c
Since some of my test cases require 64bit time_t, I implemented a detection of this situation and worked around it. Probably there is a more elegant solution to this, input how to improve this is appreciated.

##### Checklist

- [X] tests are added or updated

### CODE REVIEW

File: test/recipes/80-test_ca_internals.t [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+            }
+        }
+    }
+
+    if ($skipped) {
+        $amtexpired = $amtexpectedexpired;
+        $expirelistcorrect = 1;
+    }
+    is($exit, 1, "ca_internals_test: returned EXIT_FAILURE (".$opts->{description}.")");
+    is($amtexpired, $amtexpectedexpired, "ca_internals_test: amount of expired certificated differs from expected amount (".$opts->{description}.")");
```
Comment (mattcaswell) [inline test/recipes/80-test_ca_internals.t:163]: Typo: certificated?

File: crypto/asn1/a_time.c [hunk 2]
```diff
@@ -589,3 +589,42 @@ int ASN1_TIME_compare(const ASN1_TIME *a, const ASN1_TIME *b)
...
+    time_t timestamp_utc;
+
+    timestamp_asn1 = ASN1_TIME_new();
+    if (!ASN1_TIME_set_string(timestamp_asn1, asn1_string))
+    {
+        ASN1_TIME_free(timestamp_asn1);
+        return -1;
+    }
+
+    timestamp_tm = malloc(sizeof(*timestamp_tm));
```
Comment (mattcaswell) [inline crypto/asn1/a_time.c:614]: Use `CRYPTO_malloc`

File: crypto/asn1/a_time.c [hunk 3]
```diff
@@ -589,3 +589,42 @@ int ASN1_TIME_compare(const ASN1_TIME *a, const ASN1_TIME *b)
...
+
+    if (!(ASN1_TIME_to_tm(timestamp_asn1, timestamp_tm))) {
+        free(timestamp_tm);
```
Comment (mattcaswell) [inline crypto/asn1/a_time.c:617]: OPENSSL_free

File: crypto/asn1/a_time.c [hunk 4]
```diff
@@ -589,3 +589,42 @@ int ASN1_TIME_compare(const ASN1_TIME *a, const ASN1_TIME *b)
...
+        ASN1_TIME_free(timestamp_asn1);
+        return -1;
+    }
+
+    timestamp_local = mktime(timestamp_tm);
+    free(timestamp_tm);
```
Comment (mattcaswell) [inline crypto/asn1/a_time.c:623]: OPENSSL_free

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: test/ca_internals_test.c [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    bio_tmp = bio_err;
+    bio_err = bio_out;
+    rv = do_updatedb(db, &testdateutc);
+    bio_err = bio_tmp;
+
+    if (rv > 0) {
+        if (!save_index(indexfile, "new", db))
+            goto end;
+
+        if (!rotate_index(indexfile, "new", "old"))
```
[the remark this answers is not in the document]
Comment (t8m) [reply test/ca_internals_test.c:71]: The thing Matt is asking is to just replace the bare ! with !TEST_true(...)

[the remark this answers is not in the document]
Comment (arminfuerst) [reply test/ca_internals_test.c:71]: Thanks, I didn't get that...

[the remark this answers is not in the document]
Comment (mattcaswell) [reply test/ca_internals_test.c:71]: As per @t8m's comments simply using `!TEST_true(...)` would be sufficient to address my comment.

File: test/ca_internals_test.c [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        TEST_error("       testdate format: ASN1-String\n");
+        return 0;
+    }
+
+    /* if the test will only work with 64bit time_t and
+       the build only supports 32, assume the test as sucess */
+    need64bit = (int)strtol(test_get_argument(3), NULL, 0);
+    have64bit = sizeof(time_t) > sizeof(uint32_t);
+    if (need64bit && !have64bit) {
+        BIO_printf(bio_out, "skipping test (need64bit: %i, have64bit: %i)",
```
[the remark this answers is not in the document]
Comment (mattcaswell) [reply test/ca_internals_test.c:42]: Ah...now I see. So you are redirecting `bio_err` to `bio_out` when calling `do_updatedb`, and then printing the skip message here to `bio_out` instead if appropriate. Then the test recipes just greps the output and checks for the skip message. I can't think of a better way - so ok just leave it. Although, as a separate issue, I wonder whether just modifing `do_updatedb` to take a BIO parameter would be better than doing the dance of assigning `bio_err` to `bio_tmp`.

[the remark this answers is not in the document]
Comment (arminfuerst) [reply test/ca_internals_test.c:42]: There already was some strange ```bio_``` redirection in the code - I never understood why...
So I leave the ```BIO_printf``` and change the return to ```return TEST_skip(...)```? Or shall I also leave ```return 0```?

File: crypto/asn1/a_time.c (hunk 2 above)
Comment (arminfuerst) [reply crypto/asn1/a_time.c:614]: I am confused by the additional parameters. I changed it to ```CRYPTO_malloc(sizeof(*timestamp_tm), NULL, 0)``` and it seems to be working. Was that what you were expecting?

Comment (mattcaswell) [reply crypto/asn1/a_time.c:614]: Oops. sorry. My mistake. I actually meant `OPENSSL_malloc` (which is just a macro for `CRYPTO_malloc`, but without the extra params).

Comment (arminfuerst) [reply crypto/asn1/a_time.c:614]: Thanks, now I'm not confused anymore. :)

**Review by mattcaswell — APPROVED** [review]
LGTM

**Review by t8m — APPROVED** [review]
This is really nice work!

File: crypto/asn1/a_time.c [hunk 7]
```diff
@@ -589,3 +589,42 @@ int ASN1_TIME_compare(const ASN1_TIME *a, const ASN1_TIME *b)
...
 }
+
+/*
+ * tweak for Windows
+ */
+#ifdef WIN32
+# define timezone _timezone
+#endif
+
+time_t asn1_string_to_time_t(char *asn1_string)
```
Comment (FdaSilvaYY) [inline crypto/asn1/a_time.c:600]: Input Argument could be const ?

File: test/ca_internals_test.c [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    testdate = test_get_argument(2);
+    testdateutc = asn1_string_to_time_t(testdate);
+    if (TEST_time_t_lt(testdateutc, 0)) {
+        return 0;
+    }
+
+    indexfile = test_get_argument(1);
+    db = load_index(indexfile, NULL);
+    if (TEST_ptr_null(db)) {
+        free(indexfile);
```
Comment (FdaSilvaYY) [inline test/ca_internals_test.c:58]: are you sure that you should free an command line argument here ?

**Review by FdaSilvaYY — COMMENTED** [review]
nice work altogether

File: test/asn1_time_test.c [hunk 9]
```diff
@@ -414,5 +476,6 @@ int setup_tests(void)
     }
     ADD_ALL_TESTS(test_table_compare, OSSL_NELEM(tbl_compare_testdata));
     ADD_TEST(test_time_dup);
+	ADD_ALL_TESTS(convert_asn1_to_time_t, OSSL_NELEM(asn1_to_utc));
```
[the remark this answers is not in the document]
Comment (arminfuerst) [reply test/asn1_time_test.c:479]: the old "tab vs. spaces" :)

File: test/ca_internals_test.c (hunk 8 above)
Comment (arminfuerst) [reply test/ca_internals_test.c:58]: To be honest: no :)
I wrote this code nearly one year ago and I can't explain, why I called it here but not when checking for command line argument 2.
Same as for comment above: Shall I change it?

File: test/ca_internals_test.c [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+static int test_do_updatedb(void)
+{
+    CA_DB *db = NULL;
+    time_t testdateutc;
+    int rv;
+    size_t argc = test_get_argument_count();
+    BIO *bio_tmp;
+    char *testdate;
+    char *indexfile;
+	int need64bit;
```
[the remark this answers is not in the document]
Comment (t8m) [reply test/ca_internals_test.c:28]: Actually tab should be converted to 4 spaces

File: test/ca_internals_test.c (hunk 8 above)
Comment (t8m) [reply test/ca_internals_test.c:58]: Yes please

**Comment by arminfuerst** [comment]
One question: In the actions running in my repo (https://github.com/arminfuerst/openssl/), one build (cross-build powerpc64le-linux-gnu, libc6-dev-ppc64el-cross, linux-ppc64le) always fails - I attach the raw log. Do you have any hint what might cause this?
[powerpc64le.log](https://github.com/openssl/openssl/files/8047215/powerpc64le.log)

**Comment by t8m** [comment]
> One question: In the actions running in my repo (https://github.com/arminfuerst/openssl/), one build (cross-build powerpc64le-linux-gnu, libc6-dev-ppc64el-cross, linux-ppc64le) always fails - I attach the raw log. Do you have any hint what might cause this? [powerpc64le.log](https://github.com/openssl/openssl/files/8047215/powerpc64le.log)

This would be fixed if you rebased your branch against the current master. However please do not do this now.

**Review by t8m — APPROVED** [review]

**Review by mattcaswell — APPROVED** [review]

**Comment by t8m** [comment]
Merged after squashing the commits and fixing end-of-line whitespace issues. Thank you for your contribution!

**Pull request closed without merging**